### DISCUSSION TITLE
Differences from `mysql2` and `mysql` & roadmaps for `mysql2` in 2023

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [sidorares/node-mysql2#1841] [Q&A]
Hi @sidorares, 
I was curious about the direction of `mysql2` and where it stands in 2023.

### Question(2)
**1. What features does `mysql2` have that `mysql` does not in 2023? is the list on the homepage still accurate?**

>_[mysql2's README](https://github.com/sidorares/node-mysql2):_ 
> _Faster / Better Performance
[Prepared Statements](https://github.com/sidorares/node-mysql2/tree/master/documentation/Prepared-Statements.md)
MySQL Binary Log Protocol
[MySQL Server](https://github.com/sidorares/node-mysql2/tree/master/documentation/MySQL-Server.md)
Extended support for Encoding and Collation
[Promise Wrapper](https://github.com/sidorares/node-mysql2/tree/master/documentation/Promise-Wrapper.md)
Compression
SSL and [Authentication Switch](https://github.com/sidorares/node-mysql2/tree/master/documentation/Authentication-Switch.md)
[Custom Streams](https://github.com/sidorares/node-mysql2/tree/master/documentation/Extras.md)
[Pooling](https://github.com/sidorares/node-mysql2#using-connection-pools)_

**2. What is the relationship between the maintainer's and contributors of `mysql` vs `mysql`?**

I saw in the README the following statement:
> _MySQL2 team is working together with [mysqljs/mysql](https://github.com/mysqljs/mysql) team to factor out shared code and move it under [mysqljs](https://github.com/mysqljs/mysql) organisation._

do the projects plan to converge or would you say timing/resourcing has made it such that `mysql2` at this point is better off operating independently?

**Comment by Pieloaf** [comment]
wondering the same myself, mysqljs/mysql hasnt received a commit to master since 2022 so makes me wonder if its still being worked on/in collaboration with mysql2

**Reply by sidorares** [reply]
Hi @Pieloaf and @cliffordfajardo ( sorry missed your message now long time ago )

I might mentioned this before, but I'll try to describe past / current state of mysqljs/mysql and mysql2

- I do have write permissions to mysqljs/mysql repo and npm but would not call myself a "co-maintainer". I help answering questions very occasionally and contributed in the past tls and compression implementation ( ported from mysql2 )
- most of the maintenance work in mysqljs/mysql is / was done by @dougwilson in past years. If anyone is unhappy about speed of changes there keep in mind that it's unpaid work and Doug is busy with other projects like express, koa and many many more.

There is no active work currently to converge the codebase but making it easy for users to go from / to mysqljs/mysql is still a very high priority. Also whenever possible we explore options to factor out common code and share it rather than use own copy, recent example - https://github.com/mysqljs/aws-ssl-profiles/

**Reply by sidorares** [reply]
@cliffordfajardo to your original list:

The list is mostly accurate except for pool ( I believe it was in mysql from the beginning so not sure why this is in the list ). Also, while promise wrappers are included in mysql2 there are few modules on npm that offer that for mysqljs/mysql. There is also some code for server side protocol, but its mostly used for unit tests

**Reply by sidorares** [reply]
And roadmap for mysql2 in 2024 @cliffordfajardo 

New features:
- allow to connect with arbitrary auth plugin instead of always using `mysql_native_password `and later switching to a new auth plugin if requested by server. Quite a lot of servers use `caching_sha2_password` auth method by default so this change would save a few network roundtrips and improve connection time
- COM_RESET and cleanup of the pool connection on return in the background
- Explicitly typed prepared statements parameters. Currently types are inferred from JS parameter type, quite often this is sup optimal
- ZSTD compression support
- Better binlog protocol support - more packet types, better documentation and examples, better interoperability with existing "realtime mysql" packages
- Better server side protocol, higher level api. The current one is "packets" based which is ok for proxies and similar applications. The client is "commands" based, I want to add a similar server api based on "command - a sequence of packets state machine" abstraction.

Plus usual suspects - better performance, security and documentation, bug fixes and ideally road to 0 open issues and PRs

**Reply by dougwilson** [reply]
Hi @sidorares 👋 I will add that recently I have actually dropped most other open source work, those that have a bunch of other people on now like Express. This is because indeed I was stretched too thin especially with added on life stuff. But the future is bright and I have been off loading a lot and `mysql` module is actually on my new plan list these days to bring more up to date and work on this year without all the other distractions.

I am continuing to watch `mysql2` and I don't see `mysql2` stopping anytime soon, especially with your roadmap on it. Even with the revitalization I plan for `mysql` it would still not change the status quo that `mysql2` is a more advanced and extra features compared to `mysql`. I do plan to split apart some of the things like all that pool stuff in the old idea of it being shared too.
